Content.Shared/TapeRecorder/SharedTapeRecorderSystem.cs

```diff
@@ -0,0 +1,574 @@
+                    break;
+                case TapeRecorderMode.Stopped:
+                    args.PushMarkup(Loc.GetString("tape-recorder-stopped"));
+                    break;
+                case TapeRecorderMode.Recording:
+                    args.PushMarkup(Loc.GetString("tape-recorder-recording"));
+                    break;
+                case TapeRecorderMode.Rewinding:
+                    args.PushMarkup(Loc.GetString("tape-recorder-rewinding"));
+                    break;
+                case TapeRecorderMode.Empty:
+                    args.PushMarkup(Loc.GetString("tape-recorder-empty"));
+                    break;
+            }
+
+            //Check if we have a tape cassette inserted
+            var cassette = _itemSlotsSystem.GetItemOrNull(tapeRecorder, SlotName);
+            if (!cassette.HasValue)
+                return;
+
+            if (!TryComp<TapeCassetteComponent>(cassette, out var tapeCassetteComponent))
+                return;
+
+            OnTapeExamined((cassette.Value, tapeCassetteComponent), ref args);
+        }
+    }
+
+    protected void OnCassetteRemoved(Entity<TapeRecorderComponent> tapeRecorder, ref EntRemovedFromContainerMessage args)
+    {
+        switch (tapeRecorder.Comp.Mode)
+        {
+            case TapeRecorderMode.Playing:
+                StopPlayback(tapeRecorder);
+                break;
+            case TapeRecorderMode.Recording:
+                StopRecording(tapeRecorder);
+                break;
+            case TapeRecorderMode.Rewinding:
+                StopRewinding(tapeRecorder);
+                break;
+        }
+
+        SetMode(tapeRecorder, TapeRecorderMode.Empty);
+    }
+
+    protected void OnCassetteInserted(Entity<TapeRecorderComponent> tapeRecorder, ref EntInsertedIntoContainerMessage args)
+    {
+        SetMode(tapeRecorder, TapeRecorderMode.Stopped);
+    }
+
+    /// <summary>
+    /// Update the appearance of the tape recorder, optionally ignoring the components mode
+    /// </summary>
+    /// <param name="tapeRecorder">The tape recorder to update</param>
+    /// <param name="modeOverride">If set, use this mode instead of the mode on the component</param>
+    protected void UpdateAppearance(Entity<TapeRecorderComponent> tapeRecorder, TapeRecorderMode? modeOverride = null)
+    {
+        _appearanceSystem.SetData(tapeRecorder, TapeRecorderVisuals.Status, modeOverride.HasValue ? modeOverride : tapeRecorder.Comp.Mode);
+    }
+
+    /// <summary>
+    /// Choose a random recorded entry on the cassette and replace some of the text with hashes
+    /// </summary>
+    /// <param name="component"></param>
+    protected void CorruptRandomEntry(TapeCassetteComponent component)
+    {
+        if (component.RecordedData.Count == 0)
+            return;
+
+        var index = _robustRandom.Next(0, component.RecordedData.Count-1);
+        var entryToCorrupt = component.RecordedData[index];
+
```

Comment:
Is this supposed to not touch the last entry? This is [0, recorded count - 1)

Reply:
Ah no, i forgot that Next() is exclusive of the max bound - was trying to avoid an index out of bounds error.